### PR TITLE
update discardable size before delete files in ranges

### PR DESCRIPTION
**Pull request by Connor1996**
When using DeleteFilesInRanges of RocksDB, it would drop all SST within the ranges directly. So in this case, compaction can't be performed so that Titan knows nothing about the deleted files. 
To fix this issue, scan the properties of the to be deleted files before DeleteFilesInRange and update related blob files' discardable size to make Titan trigger GC.

### CODE REVIEW

File: src/db_impl.cc [hunk 1]
```diff
@@ -583,6 +583,145 @@ void TitanDBImpl::ReleaseSnapshot(const Snapshot* snapshot) {
...
+      std::vector<FileMetaData*> files;
+      vstorage->GetCleanInputsWithinInterval(level, begin_key, end_key, &files,
+                                             -1 /* hint_index */,
+                                             nullptr /* file_index */);
+      for (const auto& file_meta : files) {
+        if (file_meta->being_compacted) {
+          continue;
+        }
+        if (!include_end && end != nullptr &&
+            cfd->user_comparator()->Compare(file_meta->largest.user_key(),
```
Comment (yiwu-arbug) [inline src/db_impl.cc]: should this check be against file smallest key? i.e `Compare(file_meta->smallest.user_key(), *end) >= 0`.

File: src/db_impl.cc [hunk 2]
```diff
@@ -583,6 +583,145 @@ void TitanDBImpl::ReleaseSnapshot(const Snapshot* snapshot) {
...
+          continue;
+        }
+        auto fname =
+            TableFileName(cfd->ioptions()->cf_paths, file_meta->fd.GetNumber(),
+                          file_meta->fd.GetPathId());
+        if (props.count(fname) == 0) {
+          std::shared_ptr<const TableProperties> table_properties;
+          Status s =
+              version->GetTableProperties(&table_properties, file_meta, &fname);
+          if (s.ok()) {
```
Comment (yiwu-arbug) [inline src/db_impl.cc]: null check `table_properties` (despite currently rocksdb won't return null here).

File: src/titan_db_test.cc [hunk 3]
```diff
@@ -189,6 +191,32 @@ class TitanDBTest : public testing::Test {
...
+  void CompactAll() {
+    auto opts = db_->GetOptions();
+    auto compact_opts = CompactRangeOptions();
+    compact_opts.change_level = true;
+    compact_opts.target_level = opts.num_levels - 1;
+    compact_opts.bottommost_level_compaction = BottommostLevelCompaction::kSkip;
+    ASSERT_OK(db_->CompactRange(compact_opts, nullptr, nullptr));
+  }
+
+  // void IngestData(std::vector<uint64_t> data) {
```
Comment (yiwu-arbug) [inline src/titan_db_test.cc]: why comment out `IngestData`?

File: src/db_impl.cc [hunk 4]
```diff
@@ -583,6 +583,145 @@ void TitanDBImpl::ReleaseSnapshot(const Snapshot* snapshot) {
...
+  }
+
+  auto* vstorage = version->storage_info();
+  for (size_t i = 0; i < n; i++) {
+    auto begin = ranges[i].start, end = ranges[i].limit;
+    // Get all the files within range except L0, cause `DeleteFilesInRanges`
+    // would not delete the files in L0.
+    for (int level = 1; level < vstorage->num_non_empty_levels(); level++) {
+      if (vstorage->LevelFiles(i).empty() ||
+          !vstorage->OverlapInLevel(i, begin, end)) {
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/db_impl.cc]: It does, and here is actually a copy from `DeleteFilesInRanges` from RocksDB.

File: src/db_impl.cc (hunk 1 above)
Comment (Connor1996) [reply src/db_impl.cc]: seems no need, I think it is a promise of `GetCleanInputsWithinInterval`. The check here is to consider `include_end`.

File: src/db_impl.cc [hunk 5]
```diff
@@ -583,6 +583,145 @@ void TitanDBImpl::ReleaseSnapshot(const Snapshot* snapshot) {
...
 
+Status TitanDBImpl::DeleteFilesInRanges(ColumnFamilyHandle* column_family,
+                                        const RangePtr* ranges, size_t n,
+                                        bool include_end) {
+  TablePropertiesCollection props;
+  auto cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(column_family);
+  auto cfd = cfh->cfd();
+  Version* version;
+
+  // Increment the ref count
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/db_impl.cc]: Yes, it is better to have a way to get files that `DeleteFilesInRanges` delete.

[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply src/db_impl.cc]: How about send a PR to rocksdb for the change? :)

[the remark this answers is not in the document]
Comment (Connor1996) [reply src/db_impl.cc]: willing to do 😀

[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply src/db_impl.cc]: Where's the TODO comment?

File: src/db_impl.cc (hunk 1 above)
Comment (yiwu-arbug) [reply src/db_impl.cc]: ok, seems this is the same as `rocksdb::DeleteFilesInRange`

**Pull request merged**